### ISSUE TITLE
Introducing variable scopes

### ISSUE THREAD

**Issue by saig0** [zeebe-io/zeebe#1911]
Since we moved to a variables concept, we should introduce a variable scope to make it more explicitly in which scope a variable is set or can be set. The variable scope key should be used when updating variables or getting variables of a given element. The variable scope key is the element instance key if the element spawns a new variable scope (i.e. in case it can have input or output mappings). Otherwise, it is the flow scope key.

AT:
* a new variable scope is created when processing `ELEMENT_READY` or `EVENT_ACTIVATING`
* the variable scope is removed when processing `ELEMENT_COMPLETED` or `EVENT_TRIGGERED` or `ELEMENT_TERMINATED`
* add property "variableScopeKey" to workflow instance and incident record
* workflow instance record: rename property "scopeInstanceKey" to "flowScopeKey"
* variable record: rename "scopeKey" to "variableScopeKey"

**Comment by saig0** [comment]
We should add the variable scope key also to the incident record to make it easier to find and update the variables. Since the variables can be set in different scopes, it would be not clear where to update them.

**Comment by npepinpe** [comment]
Since renaming the scopeInstanceKey to something else will create a lot of noise, I've split this into a separate issue #1920 to simplify the review of the PR.

**Comment by npepinpe** [comment]
After #1939 this issue can be narrowed down to:

- Add `variableScopeKey` property to the `IncidentRecord`
- Set this key to the scope that should updated to resolve an incident; typically the element instance itself, except for incidents raised in the `ELEMENT_READY` phase of an activity due to a boundary event (e.g. correlation key extraction failed), where the `variableScopeKey` should be the scope of the activity as well

The last point might require us to first implement that correlation keys are extracted from the correct variable scope (e.g. the scope of the activity), and then we should ensure that the same scope is used to resolve a related incident.

**Comment by npepinpe** [comment]
Blocked by #1939
